pkg/block/index.go

```diff
@@ -157,45 +158,94 @@ func VerifyIndex(fn string, minTime int64, maxTime int64) error {
+	}
 	return nil
 }
 
-type IndexIssueStats struct {
-	Total int
+// CriticalErr returns error if stats indicates critical block issue, that might solved only by manual repair procedure.
+func (i Stats) CriticalErr() error {
+	var errMsg string
 
-	OutOfOrderCount int
-	OutOfOrderSum   int
-	ExactSum        int
+	if i.OutOfOrderSeries > 0 {
+		errMsg += fmt.Sprintf(
+			"%d/%d series have an average of %.3f out-of-order chunks. "+
+				"%.3f of these are exact duplicates (in terms of data and time range).",
+			i.OutOfOrderSeries,
+			i.TotalSeries,
+			float64(i.OutOfOrderChunks)/float64(i.OutOfOrderSeries),
+			float64(i.DuplicatedChunks)/float64(i.OutOfOrderChunks),
+		)
+	}
+
+	n := i.OutsideChunks - (i.CompleteOutsideChunks + i.Issue347OutsideChunks)
+	if n > 0 {
+		errMsg += fmt.Sprintf("found %d chunks non-completely outside the block time range.", n)
+	}
+
+	if i.CompleteOutsideChunks > 0 {
+		errMsg += fmt.Sprintf("found %d chunks completely outside the block time range.", i.CompleteOutsideChunks)
+	}
+
+	if errMsg != "" {
+		return errors.New(errMsg)
+	}
 
-	// Chunks that are before or after time range in meta.
-	Outsiders int
-	// Outsiders that will be never accessed. They are completely out of time range specified in block meta.
-	CompleteOutsiders int
+	return nil
 }
 
-func (i IndexIssueStats) ErrSummary() error {
-	if i.OutOfOrderCount > 0 {
-		return errors.Errorf("%d/%d series have an average of %.3f out-of-order chunks. "+
-			"%.3f of these are exact duplicates (in terms of data and time range). Outsiders: %d, complete outsiders: %d",
-			i.OutOfOrderCount, i.Total, float64(i.OutOfOrderSum)/float64(i.OutOfOrderCount),
-			float64(i.ExactSum)/float64(i.OutOfOrderSum), i.Outsiders, i.CompleteOutsiders)
+// AnyErr returns error if stats indicates any block issue.
+func (i Stats) AnyErr() error {
+	var errMsg string
+
+	if err := i.CriticalErr(); err != nil {
+		errMsg += err.Error()
+	}
+
+	if err := i.Issue347OutsideChunksErr(); err != nil {
+		errMsg += err.Error()
 	}
 
-	if i.Outsiders > 0 {
-		return errors.Errorf("No chunks are out of order, but found some outsider blocks. (Blocks that outside of block time range): %d. Complete: %d",
-			i.Outsiders, i.CompleteOutsiders)
+	if errMsg != "" {
+		return errors.New(errMsg)
 	}
 
 	return nil
 }
 
-// GatherIndexIssueStats returns useful counters as well as outsider chunks (chunks outside of block time range) that helps to assess index health.
-func GatherIndexIssueStats(fn string, minTime int64, maxTime int64) (stats IndexIssueStats, err error) {
+// GatherIndexIssueStats returns useful counters as well as outsider chunks (chunks outside of block time range) that
+// helps to assess index health.
+// It considers https://github.com/prometheus/tsdb/issues/347 as something that Thanos can handle.
+// See Stats.AcceptableOutsiders for details.
```

Comment:
I guess you renamed `Stats.AcceptableOutsiders` to `Stats.Issue347OutsideChunks`.